arrows/mvg/camera_options.cxx

```diff
@@ -0,0 +1,518 @@
+
+  GET_VALUE(bool, optimize_focal_length);
+  GET_VALUE(bool, optimize_aspect_ratio);
+  GET_VALUE(bool, optimize_principal_point);
+  GET_VALUE(bool, optimize_skew);
+  GET_VALUE(bool, optimize_dist_k1);
+  GET_VALUE(bool, optimize_dist_k2);
+  GET_VALUE(bool, optimize_dist_k3);
+  GET_VALUE(bool, optimize_dist_p1_p2);
+  GET_VALUE(bool, optimize_dist_k4_k5_k6);
+  GET_VALUE(LensDistortionType, lens_distortion_type);
+  GET_VALUE(CameraIntrinsicShareType, camera_intrinsic_share_type);
+  GET_VALUE(double, camera_path_smoothness);
+  GET_VALUE(double, camera_forward_motion_damping);
+  GET_VALUE(double, minimum_hfov);
+#undef GET_VALUE
+}
+
+/// Return true if any options to optimize intrinsic parameters are set
+bool
+camera_options
+::optimize_intrinsics() const
+{
+  if( optimize_focal_length ||
+      optimize_aspect_ratio ||
+      optimize_principal_point ||
+      optimize_skew )
+  {
+    return true;
+  }
+  switch( lens_distortion_type )
+  {
+    case POLYNOMIAL_RADIAL_DISTORTION:
+      return optimize_dist_k1
+	  || optimize_dist_k2
+	  || optimize_dist_k3;
+    case POLYNOMIAL_RADIAL_TANGENTIAL_DISTORTION:
+      return optimize_dist_k1
+	  || optimize_dist_k2
+	  || optimize_dist_k3
+	  || optimize_dist_p1_p2;
+    case RATIONAL_RADIAL_TANGENTIAL_DISTORTION:
+      return optimize_dist_k1
+	  || optimize_dist_k2
+	  || optimize_dist_k3
+	  || optimize_dist_p1_p2
+	  || optimize_dist_k4_k5_k6;
+    default:
+      break;
+  }
+  return false;
+}
+
+/// enumerate the intrinsics held constant
+std::vector<int>
+camera_options
+::enumerate_constant_intrinsics() const
+{
+  std::vector<int> constant_intrinsics;
+
+  // number of lens distortion parameters in the selected model
+  const unsigned int ndp = num_distortion_params(this->lens_distortion_type);
+
+  if (!this->optimize_focal_length)
+  {
+    constant_intrinsics.push_back(0);
+  }
+  if (!this->optimize_principal_point)
+  {
+    constant_intrinsics.push_back(1);
+    constant_intrinsics.push_back(2);
+  }
+  if (!this->optimize_aspect_ratio)
+  {
+    constant_intrinsics.push_back(3);
+  }
+  if (!this->optimize_skew)
+  {
+    constant_intrinsics.push_back(4);
+  }
+  if (!this->optimize_dist_k1 && ndp > 0)
+  {
+    constant_intrinsics.push_back(5);
+  }
+  if (!this->optimize_dist_k2 && ndp > 1)
+  {
+    constant_intrinsics.push_back(6);
+  }
+  if (!this->optimize_dist_p1_p2 && ndp > 3)
+  {
+    constant_intrinsics.push_back(7);
+    constant_intrinsics.push_back(8);
+  }
+  if (!this->optimize_dist_k3 && ndp > 4)
+  {
+    constant_intrinsics.push_back(9);
+  }
+  if (!this->optimize_dist_k4_k5_k6 && ndp > 7)
+  {
+    constant_intrinsics.push_back(10);
+    constant_intrinsics.push_back(11);
+    constant_intrinsics.push_back(12);
+  }
+  return constant_intrinsics;
+}
+
+/// extract the extrinsic paramters from a camera into the parameter array
+void
+camera_options
+::extract_camera_extrinsics(const camera_perspective_sptr camera,
```

Comment:
All of the remaining functions in this file that convert camera models to or from raw arrays of doubles are specific to the use in Ceres.  I would put those functions back in Ceres.

Reply:
moved them back to ceres